pgcontents/tests/test_synchronization.py

```diff
@@ -177,3 +184,99 @@ def check_reencryption(old, new):
+        managers = {}
+        for user_id in user_ids:
+            managers[user_id] = PostgresContentsManager(
+                user_id=user_id,
+                db_url=db_url,
+                crypto=crypto_factory(user_id),
+                create_user_on_startup=True,
+            )
+        paths = {}
+
+        def create_nbs(user_id):
+            """
+            Create test notebooks and create a checkpoint for each.
+            """
+            paths[user_id] = set(populate(managers[user_id]))
+            for path in paths[user_id]:
+                managers[user_id].create_checkpoint(path)
+
+        # Create test files/remote checkpoints for both users, their creation
+        # times separated by a recorded datetime.
+        create_nbs(user_ids[0])
+        split_dt = datetime.now()
+        create_nbs(user_ids[1])
+
+        def process_files_nb(nb):
+            """
+            For notebook from files table, creates path by concatenating parent
+            and filename.
+            """
+            nb['path'] = nb['parent_name'] + nb['name']
+
+        def check_generate_call(generate_fn, expect_users, kwargs={}):
+            """
+            Calls the given function and checks that all notebooks belonging to
+            the specified users are found.
+            """
+            path_record = {}
+            for user_id in expect_users:
+                path_record[user_id] = set()
+            for result in generate_fn(engine, crypto_factory, **kwargs):
+                if generate_fn is generate_files:
+                    process_files_nb(result)
+                self.assertIn(result['user_id'], expect_users)
+
+                # Converts the content result to a dict format matching the
+                # return value of `PostgresContentManager.get()`.
+                nb = reads(result['content'], as_version=NBFORMAT_VERSION)
+                managers[result['user_id']].mark_trusted_cells(nb,
+                                                               result['path'])
+                self.assertEqual(
+                    managers[result['user_id']].get(result['path'])['content'],
+                    nb
+                )
+                path_record[result['user_id']].add(result['path'][1:])
+
+            # Makes sure all notebooks were found.
+            for user_id in expect_users:
+                self.assertEqual(path_record[user_id], paths[user_id])
+
+        def check_generate_fn(generate_fn):
+            """
+            Checks the given function with different kwarg calls.
+            """
+            max_kwargs = {'max_dt': split_dt}
+            min_kwargs = {'min_dt': split_dt}
+
+            check_generate_call(generate_fn, user_ids)
+            # User 0's notebooks were created before `split_dt`, so they'll be
+            # found by the `max_dt` call; the opposite applies to User 1's
+            # notebooks.
+            check_generate_call(generate_fn, [user_ids[0]], max_kwargs)
```

Comment:
Also, querying for notebooks with more than one checkpoint, which probably requires checkpoint-specific logic.  My inclination would probably be to separate the tests for notebooks and checkpoints rather than trying to test both at the same time.

As a general philosophical point, I'm much more likely to prefer a modest amount of code duplication over building a shared abstraction when writing tests, because the most important thing about a test is that it should be clear to the reader what behavior is under test and why the assertions being made follow from the configuration under test.